File: Android/app/src/main/java/com/pokeapi/lpiem/pokeapiandroid/Provider/RetrofitSingleton.java
package com.pokeapi.lpiem.pokeapiandroid.Provider;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.pokeapi.lpiem.pokeapiandroid.Model.Pokemon.PokemonAPI;

import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitSingleton {
    private static PokemonAPI instance = null;
    public final static String POKEMONBASEURL = "https://pokeapi.co/";

    public static PokemonAPI getInstance(){
        if(instance == null){
            instance = buildInstance();
        }
        return instance;
    }

    private static PokemonAPI buildInstance(){
        Gson gson = new GsonBuilder().setLenient().create();
        Retrofit retrofit = new Retrofit.Builder().baseUrl(POKEMONBASEURL).addConverterFactory(GsonConverterFactory.create(gson)).build();
        PokemonAPI pokemonAPI = retrofit.create(PokemonAPI.class);
        return pokemonAPI;
    }

    private RetrofitSingleton(){

    }
}

File: Android/app/src/main/java/com/pokeapi/lpiem/pokeapiandroid/provider/RetrofitSingleton.kt
package com.pokeapi.lpiem.pokeapiandroid.provider

import android.util.Log
import com.google.gson.GsonBuilder
import com.pokeapi.lpiem.pokeapiandroid.Model.Retrofit.Pokemons.PokemonAPI
import com.pokeapi.lpiem.pokeapiandroid.Model.Retrofit.Pokemons.PokemonList
import com.pokeapi.lpiem.pokeapiandroid.R
import com.pokeapi.lpiem.pokeapiandroid.View.Fragment.PokedexListView
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import com.pokeapi.lpiem.pokeapiandroid.model.retrofit.pokemons.PokemonAPI

import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

const val CUSTOMPOKEAPIBASEURL = "https://walkemon.herokuapp.com/"

object RetrofitSingleton {
    /**
     * Build an instance of the pokemonAPI which to Run
     * @return instance of pokemonAPI retrofit routes in order to fetch data from the api
     */
    fun buildInstance(): PokemonAPI {
        val gson = GsonBuilder().setLenient().create()
        val retrofit = Retrofit.Builder().baseUrl(CUSTOMPOKEAPIBASEURL).addConverterFactory(GsonConverterFactory.create(gson)).build()
        return retrofit.create(PokemonAPI::class.java)
    }
}
